### ISSUE TITLE
Why does GlobPtr depend on the pattern type

### ISSUE THREAD

**Issue by fmoessbauer** [dash-project/dash#325]
The `dash::GlobPtr` has to be declared with a pattern type, but I cannot find a single method which requires this.

```cpp
template<
  typename ElementType,
  class    PatternType  = Pattern<1> >
class GlobPtr {...};
```

**Comment by fuchsto** [comment]
This is a good one, and it raises almost philosophical concerns.
The reason for the dependency on a pattern is the possibility to define pointer arithmetics corresponding to iteration space.

The ancient question behind this is the difference between `GlobPtr` and `GlobIter`.
With `GlobRef` it is easy because references cannot be re-directed, just like vanilla reference types.

Now assuming a global pointer to an element in a matrix block where block elements are non-contiguous (e.g. BlockPattern, opposed to TilePattern). When incrementing the global pointer, should it traverse elements in iteration order like `GlobIter` or not?

My suggestion is that it should not. Just like incrementing a pointer to an element of `std::list`, it would not refer to the next element in the list as pointers are agnostic of iteration order.
Pointers are dull, non-validated value-representations of addresses, iterators respect the structure of values in memory.

A often-cited formal concern is that pointers *are* iterators, or rather: iterators are a more general concept of pointers, so they should behave the same. In my understanding, `GlobPtr` is the PGAS equivalent of `T *`, so with a nod in the direction of category theory, we should be fine without the pattern dependency.

@dash-project/developers Everyone, let me know if you have objections on this.

**Comment by fuerlinger** [comment]
> My suggestion is that it should not. Just like incrementing a pointer to an element of std::list, it would not refer to the next element in the list as pointers are agnostic of iteration order.
Pointers are dull, non-validated value-representations of addresses, iterators respect the structure of values in memory.

:+1:

**Comment by devreal** [comment]
Agreed. It is up to the user to be careful when juggling with (global) pointers, or just don't do it. We should document though that a `dash::Array` is different from a `std::array` in that you cannot iterate over the whole array by continuously incrementing a `GlobPtr` as you would do with a `T *` and a `std::array`. 

Just one remark: The definition of `self_t operator+=(gptrdiff_t n)` in `GlobPtr` seems wrong, it should rather be returning a reference to itself. The same holds true for `operator-=`.

**Comment by fuchsto** [comment]
@devreal The semantic deviation of `GlobPtr` from `T *`, which also affects global pointer distance, is bugging me since I started working on DASH. The proposal in PR #333 tries to fix it.

**Comment by fuchsto** [comment]
Fixed in PR #333.
